src/mscorlib/src/System/Text/StringBuilder.CoreCLR.cs

```diff
@@ -8,11 +8,94 @@ namespace System.Text
+            {
+                newCapacity = (requiredCapacity + 1) & ~1;
+            }
+            return newCapacity;
+        }
+
+        internal unsafe void ReplaceBufferInternal(char* newBuffer, int newLength)
+        {
+            if (newLength > m_MaxCapacity)
+                throw new ArgumentOutOfRangeException("capacity", SR.ArgumentOutOfRange_Capacity);
+
+            // Leave space for a NUL character at the end, the previous implementation did this
+            m_ChunkChars = new char[GetReplaceBufferCapacity(newLength + 1)];
+
+            fixed (char* pChunkChars = m_ChunkChars)
+            {
+                Buffer.MemoryCopy(newBuffer, pChunkChars, m_ChunkChars.Length * sizeof(char), newLength * sizeof(char));
+            }
+
+            m_ChunkLength = newLength;
+            m_ChunkPrevious = null;
+            m_ChunkOffset = 0;
+        }
 
-        [MethodImplAttribute(MethodImplOptions.InternalCall)]
-        internal extern unsafe void ReplaceBufferAnsiInternal(sbyte* newBuffer, int newLength);
+        internal unsafe void ReplaceBufferAnsiInternal(sbyte* newBuffer, int newLength)
+        {
+            if (newLength > m_MaxCapacity)
+                throw new ArgumentOutOfRangeException("capacity", SR.ArgumentOutOfRange_Capacity);
+
+            // Leave space for a NUL character at the end, the previous implementation did this
+            m_ChunkChars = new char[GetReplaceBufferCapacity(newLength + 1)];
+
+            int convertedChars;
+
+            fixed (char* pChunkChars = m_ChunkChars)
+            {
+                Diagnostics.Debug.Assert(newLength < m_ChunkChars.Length);
+
+                // The incoming string buffer is supposed to have been populated by the 
+                // P/Invoke-called native function but there's no way to know if that really
+                // happened, the native function might populate the buffer only in certain
+                // circumstances (e.g. only if the function succeeds).
+                //
+                // As such, the buffer might contain bogus characters that cannot be converted 
+                // to Unicode and in that case conversion should not result in exceptions that
+                // the managed caller does not expect. Instead, the caller is expected to know
+                // when the resulting string is not valid and not use it.
+                //
+                // Both MultiByteToWideChar and the UTF8Encoding instance used on Unix-like
+                // platforms default to replacing invalid characters with the Unicode replacement
+                // character U+FFFD.
+#if PLATFORM_UNIX
+                try
+                {
+                    convertedChars = Encoding.UTF8.GetChars((byte*)newBuffer, newLength, pChunkChars, newLength);
+                }
+                catch (ArgumentException)
+                {
+                    // This should not normally happen because converting N UTF-8 bytes to UTF-16
```

Comment:
I would not bother with the `try/catch` if this is something that should never happen.

Reply:
Well, trouble is I'm only 99.99% sure it can't happen so I kept it there to be sure. `UTF8Encoding.GetMaxCharCount` returns the number of bytes + 1:
https://github.com/dotnet/coreclr/blob/466da32a4b47e63b4690f78d4cf87ac3fbf67487/src/mscorlib/shared/System/Text/UTF8Encoding.cs#L2445-L2453
and the explanation `if 1st byte is last byte of 4 byte surrogate pair` is not terrible clear. Surrogate pairs are a UTF-16 thing so what exactly a 4 byte surrogate pair is supposed to be in UTF-8!?

Reply:
Best I can tell from `GetCharCount` code, the `GetMaxCharCount` comment only applies only to `Decoder.GetChars`. In that case you can have state from a previous call - decoding the last byte of a 4 byte UTF-8 sequence will result in 2 UTF-16 surrogates being generated. For `Encoding.GetChars` this doesn't apply, that last byte is a trailing byte that will be treated as an invalid byte and generate a single fallback character.

After all, if a magic byte sequence existed that could generate more than one character per byte, such a sequence could occur multiple times and then `GetMaxCharCount` couldn't simply add 1.

So I'll drop the try/catch.